src/main/java/coresearch/cvurl/io/request/CVurlProxySelector.java

```diff
@@ -0,0 +1,79 @@
+package coresearch.cvurl.io.request;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class CVurlProxySelector extends ProxySelector {
+
+    private final Map<String, List<Proxy>> proxiesByUri = new ConcurrentHashMap<>();
+    private final ProxySelector proxySelector;
+
+    public CVurlProxySelector() {
+        this(null);
+    }
+
+    public CVurlProxySelector(ProxySelector proxySelector) {
+        this.proxySelector = proxySelector;
+    }
+
+    @Override
+    public List<Proxy> select(URI uri) {
+        // proxy-per-request
+        final List<Proxy> proxies = new ArrayList<>(proxiesByUri.getOrDefault(uri.toString(), List.of()));
+
+        // proxy-per-client
+        if (proxies.isEmpty() && proxySelector != null)
+            proxies.addAll(proxySelector.select(uri));
+
+        // default proxy
+        if (proxies.isEmpty())
+            proxies.addAll(ProxySelector.getDefault().select(uri));
+
+        return proxies;
+    }
+
+    @Override
+    public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
+        // ignore for now
+    }
+
+    public void addProxy(String uri, CVurlProxy cVurlProxy) {
+        proxiesByUri.computeIfAbsent(uri, k -> new ArrayList<>())
+                .add(toProxy(cVurlProxy));
+    }
+
+    public void removeProxiesForUri(URI uri) {
```

Comment:
why do we want to do this?

Reply:
Because `HttpClient` holds an instance of `CVurlProxySelector` and so if we won't remove proxies after the request we will use same proxies for subsequent calls to this URI even if user will not use `RequestBuilder#withProxy` method explicitly. 

        final Response<String> firstResponse = cvurl.get("https://google.com/")
                .withProxy(CVurlProxy.of(CVurlProxyType.HTTP, "proxy.host", 8080))
                .asString()
                .orElse(null);
        final Response<String> secondResponse = cvurl.get("https://google.com/")
                .asString()
                .orElse(null);

In example above, second request to the same URI will use proxy from first request. Method `removeProxiesForUri` prevents this behavior.